devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/pom/PomUtils.java

```diff
@@ -0,0 +1,69 @@
+package io.quarkus.gradle.tooling.pom;
+
+import java.io.FileReader;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Model;
+import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.result.ArtifactResolutionResult;
+import org.gradle.api.artifacts.result.ArtifactResult;
+import org.gradle.api.artifacts.result.ComponentArtifactsResult;
+import org.gradle.api.internal.artifacts.DefaultModuleIdentifier;
+import org.gradle.api.internal.artifacts.result.DefaultResolvedArtifactResult;
+import org.gradle.internal.component.external.model.DefaultModuleComponentIdentifier;
+import org.gradle.maven.MavenModule;
+import org.gradle.maven.MavenPomArtifact;
+
+import io.quarkus.maven.dependency.ArtifactCoords;
+
+public final class PomUtils {
+    public static Set<String> readDeploymentDependencies(Project project, ArtifactCoords artifactCoords) {
```

Comment:
I will let @aloubyansky  chime in but I'm pretty sure we don't want to read the POM.

What I don't exactly understand is how you end up including these deployment dependencies given they shouldn't be part of the dependencies at all. But I'm not exactly sure how you get to them.

Reply:
At the moment we go recursively through all dependencies to find the extensions for which we also need to add deployment modules.

Another option would be to turn the enforcement up to 11 and only consider first level extension dependencies. But there we would lose out on extensions pulled in by normal libraries (which is a use-case I have seen, because I do it).

For the Gradle Extension plugin we'd also then need to go much harder on enforcement for the deployment module dependencies, meaning we'd have to make the validation task mandatory for building the final artifact. This would also include a change to the documentation making adding the deployment modules of other extensions mandatory, otherwise they would be missing from the deployment classpath.

Reply:
What I don't understand is that I would expect that you would have an easy way in Gradle to resolve transitive dependencies of an app, and I would expect these transitive dependencies to respect the exclusions that are in the POM files.
I don't understand why we are trying to resolve things ourselves and thus ignoring the exclusions.

Note that I'm not familiar at all with the Gradle API so I could be completely mistaken.

Reply:
I'll copy here our exchange on the original commit @vemilyus 
```
aloubyansky 

One point to keep in mind is that, when we resolve deployment dependencies, we need to resolve only the top level one deployment artifact. So in that issue we are fixing, it'll be quarkus-hibernate-reactive-deployment. It will already resolve its dependencies with the exclusions applied.
Then we "simply" replace the quarkus-hibernate-reactive with quarkus-hibernate-reactive-deployment and its dependencies (where quarkus-hibernate-reactive is also present) to create the Quarkus build classpath.

However, when we resolve quarkus-hibernate-reactive-deployment we should apply exclusions from the dependencies that appear higher in the hierarchy, e.g. if quarkus-hibernate-reactive was a dependency of acme-lib and my-app had a dependency on acme-lib with quarkus-hibernate-reactive pulled in as its transitive dependency. So, if my-app configured dependency exclusions on acme-lib dependency, those exclusions that should also apply when we resolve dependencies of quarkus-hibernate-reactive-deployment.

vemilyus  Author Owner

Let me address your concerns one at a time:

We can never guarantee that deployment modules pull in all the deployment modules of an extension's runtime dependencies. We also don't enforce the validateExtension task in the extension plugin, which can also be disabled.
By only resolving the first level artifacts we lose out on extensions pulled in by normal libraries.
Extensions that are dependencies of normal libs (acme-lib) wouldn't be recognized as extensions, thus their deployment module wouldn't be added to the deployment classpath thereby nullifying its functionality.
```

Reply:
> We can never guarantee that deployment modules pull in all the deployment modules of an extension's runtime dependencies.

Why not? The extension dependency model implies consistency between runtime and deployment dependencies. Meaning if ext-a-runtime depends on ext-b-runtime then ext-a-deployment will depend on ext-b-deployment. If that rule is not followed there is not guarantee that extension will work, iow it's not supported.

Reply:
> By only resolving the first level artifacts we lose out on extensions pulled in by normal libraries.

If by first level you mean direct dependencies and in that case you are correct. Which is why we walk dependency trees until we meet the top level runtime extension artifact - that one will be the "first level" from that point.

Reply:
So, if I understand you correctly we only walk regular libraries recursively and when we hit an extension we don't walk its dependencies?

Then I'd like to bring to mind my comment from two days ago where I mentioned that we should go hard on validation of extensions then, to make sure that only extensions which fit the model you mentioned can get published.

Reply:
@aloubyansky In addition I'd propose that I then also add the feature to the Gradle extension plugin to automatically add the deployment modules of runtime extension dependencies to the deployment module unless the user specifies them himself, so we can make somewhat sure the deployment dependencies aren't forgotten :grin:

Reply:
We need to visit all the dependencies and identify all the extension artifacts either way since we need to make sure we activate all the conditional dependencies that need to be activated.

But when it comes to "injecting" deployment dependencies, we essentially replace the top level runtime extension dependency node (whether it appears to be a direct or a transitive dependency of an app) with its corresponding deployment dependency node that we would resolve in the plugin.

+1 to improving the extension dependency validation check.

I wouldn't add dependencies automatically. I would prefer failing the build with an error suggesting what should be fixed in the project.

Thanks a lot for looking into this @vemilyus

Reply:
> I wouldn't add dependencies automatically. I would prefer failing the build with an error suggesting what should be fixed in the project.

Would you be okay with putting this behind a configuration value (disabled by default)? 

I'd like to have this feature and by putting it behind a flag it wouldn't change the default behavior and be explicit to the user. Ofc I can put it in a separate PR if you prefer.

Reply:
If you really like to do that, I'd separate it from the rest of the work related to fixing the issue.